cub/warp/warp_exchange.cuh

```diff
@@ -468,5 +468,89 @@ public:
   //@}  end member group
 };
 
+template <typename InputT, int ITEMS_PER_THREAD>
+class WarpExchangeRegister
+{
+  static_assert(PowerOfTwo<ITEMS_PER_THREAD>::VALUE, "ITEMS_PER_THREAD must be a power of two");
+
+  template <int SIZE, int idx>
+  class CompileTimeArrayImpl; // fwd decl for partial specializations
+
+  template <int SIZE> // terminating partial specialization
+  class CompileTimeArrayImpl<SIZE, SIZE>
+  {
+  protected:
+    static_assert(PowerOfTwo<SIZE>::VALUE, "SIZE must be a power of two");
+
+    // used for dumping back the individual values after transposing
+    InputT (&output_items)[SIZE];
+
+    template <int>
+    __device__ void Loop(bool, unsigned)
+    {}
+
+  public:
+    __device__ CompileTimeArrayImpl(InputT (&input_items)[SIZE])
+        : output_items{input_items}
+    {}
+  };
+
+  template <int SIZE, int idx> // concrete recursion class
+  class CompileTimeArrayImpl : protected CompileTimeArrayImpl<SIZE, idx + 1>
```

Comment:
This is pattern used by [Trove](https://github.com/bryancatanzaro/trove/blob/master/trove/array.h#L36) too and something I've been doing a lot in my code. In my experience, hoisting an array into registers places a major "mental load" on the optimizer, and it often then becomes a game of whack-a-mole by sprinkling `#pragma unroll` in front of various nested loops and praying that the compiler doesn't drop all of into local memory. Using "compile-time" arrays is a lot easier to optimize from the compiler's point of view, since the code never subscripts an array and only accesses values using compile-time indexes.